### PR TITLE
fix metal

### PR DESCRIPTION
**Pull request by Mee-gu**
fix metal buffer usage

### CODE REVIEW

File: cocos/renderer/gfx-gles2/GLES2Commands.cc [hunk 1]
```diff
@@ -1395,6 +1395,17 @@ void GLES2CmdFuncExecuteCmds(GLES2Device* device, GLES2CmdPackage* cmd_package)
...
+          if(cmd->viewportDirty)
+          {
+              glViewport(cmd->viewport.left, cmd->viewport.top, cmd->viewport.width, cmd->viewport.height);
+              cache->viewport = cmd->viewport;
+          }
+          if(cmd->scissorDirty)
+          {
+              glScissor(cmd->scissor.x, cmd->scissor.y, cmd->scissor.width, cmd->scissor.height);
+              cache->scissor = cmd->scissor;
+          }
```
Comment (YunHsiao) [inline cocos/renderer/gfx-gles2/GLES2Commands.cc:1407]: Is this for dynamic state usage? Should handle all the cases(stencil masks, etc.)

Could take a look at the webgl backend on this, I don't think the dirty flags are necessary

File: cocos/renderer/core/gfx/GFXDevice.h [hunk 2]
```diff
@@ -104,6 +104,8 @@ class CC_CORE_API GFXDevice : public Object {
     bool _reverseCW = false;
     uint _shaderIdGen = 0;
     std::unordered_map<String, String> _macros;
+    GFXFormat _colorFmt = GFXFormat::UNKNOWN;
+    GFXFormat _depthStencilFmt = GFXFormat::UNKNOWN;
```
Comment (minggo) [inline cocos/renderer/core/gfx/GFXDevice.h:108]: Where are they initialized? `_context` may change, for example, after returning from background on Android, GL ES  context may be recreated.

File: cocos/renderer/gfx-metal/MTLFrameBuffer.mm [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#include "MTLFrameBuffer.h"
+#include "MTLRenderPass.h"
+#include "MTLTextureView.h"
+
+NS_CC_BEGIN
+
+CCMTLFrameBuffer::CCMTLFrameBuffer(GFXDevice* device) : GFXFramebuffer(device) {}
+CCMTLFrameBuffer::~CCMTLFrameBuffer() { destroy(); }
+
+bool CCMTLFrameBuffer::initialize(const GFXFramebufferInfo& info)
```
Comment (YunHsiao) [inline cocos/renderer/gfx-metal/MTLFrameBuffer.mm:11]: Would be better if keepping the same naming convention as other backends: Framebuffer, with lowercase 'b' (it's one word)

File: cocos/renderer/gfx-gles2/GLES2Commands.cc [hunk 4]
```diff
@@ -1395,6 +1395,17 @@ void GLES2CmdFuncExecuteCmds(GLES2Device* device, GLES2CmdPackage* cmd_package)
         GLES2CmdBindStates* cmd = cmd_package->bindStatesCmds[cmd_idx];
         is_shader_changed = false;
           
+          if(cmd->viewportDirty)
+          {
+              glViewport(cmd->viewport.left, cmd->viewport.top, cmd->viewport.width, cmd->viewport.height);
+              cache->viewport = cmd->viewport;
+          }
```
Comment (minggo) [inline cocos/renderer/gfx-gles2/GLES2Commands.cc:1402]: If use like this, then does `cache` need to record viewport?

File: cocos/renderer/core/gfx/GFXDevice.h (hunk 2 above)
Comment (Mee-gu) [reply cocos/renderer/core/gfx/GFXDevice.h:108]: they are initialized in XXXDevice.xx source file

File: cocos/renderer/gfx-gles2/GLES2Commands.cc (hunk 4 above)
Comment (Mee-gu) [reply cocos/renderer/gfx-gles2/GLES2Commands.cc:1402]: begin render pass will use cache to setup viewport

File: cocos/renderer/gfx-gles2/GLES2Commands.cc (hunk 1 above)
Comment (minggo) [inline cocos/renderer/gfx-gles2/GLES2Commands.cc:1407]: The same as viewport.

**Comment by YunHsiao** [comment]
I think these changes could be submitted as a seperate PR to the main repository, since the CI isn't affected by any of the changes here

File: cocos/renderer/gfx-gles2/GLES2Commands.cc (hunk 1 above)
Comment (Mee-gu) [reply cocos/renderer/gfx-gles2/GLES2Commands.cc:1407]: not yet for dynamic state usage, just fix gfx-test-case. The dynamic state will be achieve in another PR

**Comment by minggo** [comment]
@YunHsiao 
Does it depend on your PR?

File: cocos/renderer/gfx-metal/MTLFrameBuffer.mm (hunk 3 above)
Comment (Mee-gu) [reply cocos/renderer/gfx-metal/MTLFrameBuffer.mm:11]: got it

File: cocos/renderer/gfx-gles2/GLES2Commands.cc (hunk 1 above)
Comment (YunHsiao) [reply cocos/renderer/gfx-gles2/GLES2Commands.cc:1407]: but it's part of the dynamic state functionalities right? Sorry but I don't think this is the right way to do dynamic states

**Comment by Mee-gu** [comment]
> I think these changes could be submitted as a seperate PR to the main repository, since the CI isn't affected by any of the changes here

i will take a look about the CI compile issue

**Comment by YunHsiao** [comment]
> @YunHsiao
> Does it depend on your PR?

Yes, but we could manage that by just merge mine first, the scope of this goes much beyond simple CI fix

**Comment by Mee-gu** [comment]
> @YunHsiao
> Does it depend on your PR?

I push the pr here in case of @PartriceJiang can run the gfx-test-case

File: cocos/renderer/gfx-metal/MTLBuffer.mm [hunk 5]
```diff
@@ -74,6 +70,18 @@ MTLResourceOptions toMTLResourseOption(GFXMemoryUsage usage)
...
+        _mtlResourceOptions = mu::toMTLResourseOption(info.memUsage);
+        _mtlBuffer = [id<MTLDevice>(((CCMTLDevice*)_device)->getMTLDevice() ) newBufferWithLength:_size
+                                                                                          options:_mtlResourceOptions];
+        if (_mtlBuffer == nil)
+        {
+            _status = GFXStatus::FAILED;
+            CCASSERT(false, "Failed to create MTLBuffer.");
+            return false;
+        }
+    }
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLBuffer.mm:84]: I think can use a new function as vertex and index buffer can share some logic.

File: cocos/renderer/gfx-metal/MTLBuffer.mm [hunk 6]
```diff
@@ -58,12 +49,17 @@ MTLResourceOptions toMTLResourseOption(GFXMemoryUsage usage)
     }
     
     if (_usage & GFXBufferUsage::VERTEX ||
-        _usage & GFXBufferUsage::INDEX ||
         _usage & GFXBufferUsage::UNIFORM)
     {
-        _mtlResourceOptions = toMTLResourseOption(info.memUsage);
-        if (_size > 0)
+//        for single-use data smaller than 4 KB, use setVertexBytes:length:atIndex: instead
+        if (_size < MINIMUMR_REQUIRED_SIZE_4KB)
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLBuffer.mm:55]: It is better to add the link for the optimization to describe why use small vertex buffer like this.

File: cocos/renderer/gfx-metal/MTLDevice.mm [hunk 7]
```diff
@@ -16,8 +16,10 @@
 #include "MTLTexture.h"
 #include "MTLTextureView.h"
 #include "MTLSampler.h"
-#import <MetalKit/MTKView.h>
 #include "MTLUtils.h"
+#include <platform/mac/CCView.h>
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLDevice.mm:20]: Why need to include this header file?

File: cocos/renderer/gfx-metal/MTLDevice.mm [hunk 8]
```diff
@@ -287,4 +292,36 @@
...
+                          sourceOffset:0
+                              toBuffer:id<MTLBuffer>(dstBuffer)
+                     destinationOffset:offset
+                                  size:size];
+    [blitCommandEncoder endEncoding];
+
+    // Add a completion handler and commit the command buffer.
+    [commandBuffer addCompletedHandler:^(id<MTLCommandBuffer> cb) {
+        // Private buffer is populated.
+    }];
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLDevice.mm:323]: Do nothing in complete handler, why need it?

File: cocos/renderer/gfx-metal/MTLUtils.mm [hunk 9]
```diff
@@ -231,6 +231,20 @@ EShLanguage getShaderStage(GFXShaderType type)
 
 namespace mu
 {
+    MTLResourceOptions toMTLResourseOption(GFXMemoryUsage usage)
+    {
+        if (usage & GFXMemoryUsage::HOST && usage & GFXMemoryUsage::DEVICE)
```
Comment (minggo) [inline cocos/renderer/gfx-metal/MTLUtils.mm:236]: It is better to add `()` for `&`.

**Comment by minggo** [comment]
> > @YunHsiao
> > Does it depend on your PR?
> 
> Yes, but we could manage that by just merge mine first, the scope of this goes much beyond simple CI fix

@Mee-gu yep, please send a PR to official repo after passing review.

File: cocos/renderer/gfx-metal/MTLDevice.mm (hunk 8 above)
Comment (Mee-gu) [reply cocos/renderer/gfx-metal/MTLDevice.mm:323]: to make sure source data did populated to GPU

File: cocos/renderer/gfx-metal/MTLDevice.mm (hunk 7 above)
Comment (Mee-gu) [reply cocos/renderer/gfx-metal/MTLDevice.mm:20]: in `blitBuffer`, a mtlCommandQueue is required to create the blit command buffer.

**Pull request closed without merging**